zilliqa/src/block_store.rs

```diff
@@ -82,12 +87,27 @@ impl BlockStore {
             self.highest_known_view = view;
         }
 
-        self.request_missing_blocks(random_peer)?;
+        let peer = self.peers.entry(from).or_default();
+        if view > peer.highest_known_view {
+            trace!(%from, view, "new highest known view for peer");
+            peer.highest_known_view = view;
+        }
+
+        self.request_missing_blocks()?;
 
         Ok(())
     }
 
-    pub fn request_missing_blocks(&mut self, random_peer: Option<PeerId>) -> Result<()> {
+    pub fn best_peer(&self, view: u64) -> Option<PeerId> {
+        let (best, _) = self
+            .peers
+            .iter()
+            .filter(|(_, peer)| peer.highest_known_view >= view)
+            .min_by_key(|(_, peer)| peer.requested_blocks)?;
```

Comment:
Hopefully, "most" peer.requested_blocks will be -1, which means you'll hammer whichever one min_by_key finds first. Not important, but I thought I'd say.

Reply:
Hmm good point, it would be nice if we shuffled through all equally good peers. I'll raise an issue for this.

Reply:
https://github.com/Zilliqa/zq2/issues/1133